data_safe_haven/infrastructure/stack_manager.py

```diff
@@ -1,24 +1,113 @@
 from data_safe_haven.exceptions import DataSafeHavenAzureError, DataSafeHavenPulumiError
-from data_safe_haven.external import AzureApi, AzureCli
+from data_safe_haven.external import AzureApi
 from data_safe_haven.functions import replace_separators
 from data_safe_haven.infrastructure.stacks import DeclarativeSHM, DeclarativeSRE
 from data_safe_haven.utility import LoggingSingleton
 
 
+def handle_pulumi_login(config) -> None:
+    account = PulumiAccount(config)
+
+    if not account.confirm():
+        account.login()
+        if not account.confirm():
+            raise typer.Exit()
+
+
+def pulumi_env(config: Config) -> dict[str, Any]:
+    azure_api = AzureApi(config.subscription_name)
+    backend_storage_account_keys = azure_api.get_storage_account_keys(
+        config.backend.resource_group_name,
+        config.backend.storage_account_name,
+    )
+    return {
+        "AZURE_STORAGE_ACCOUNT": config.backend.storage_account_name,
+        "AZURE_STORAGE_KEY": str(backend_storage_account_keys[0].value),
+        "AZURE_KEYVAULT_AUTH_VIA_CLI": "true",
+    }
+
+
+class PulumiAccount:
+    def __init__(self, config: Config):
+        self.cfg = config
+        self.path = which("pulumi")
+        if self.path is None:
+            msg = "Unable to find Pulumi CLI executable in your path.\nPlease ensure that Pulumi is installed"
+            raise DataSafeHavenPulumiError(msg)
+        self.env_: dict[str, Any] | None = None
+
+    @property
+    def env(self) -> dict[str, Any]:
+        """Get necessary Pulumi environment variables"""
+        if not self.env_:
+            self.env_ = pulumi_env(self.cfg)
+        return self.env_
+
+    def confirm(self) -> bool:
+        """Prompt user to confirm the Pulumi account is correct"""
+        # Because the who_am_i method requires a stack and workspace, it is difficult to
+        # do this with a minimal dummy stack which also works with the Azure backend
+        # stack = automation.create_or_select_stack(
+        #     stack_name="dummy_stack",
+        #     project_name="dummy_project",
+        #     work_dir="./",
+        #     opts=automation.LocalWorkspaceOptions(
+        #         env_vars=self.env,
+        #     ),
+        # )
+        # result = stack.workspace.who_am_i()
+        # print(
+        #     f"user: {result.user}\n",
+        #     f"url: {result.url}\n",
+        #     f"organisations: {result.organizations}"
+        # )
+        try:
+            result = subprocess.check_output(
+                [self.path, "whoami", "--verbose"],
+                stderr=subprocess.PIPE,
+                encoding="utf8",
+                env=self.env
+            )
+        except subprocess.CalledProcessError as exc:
+            msg = f"Logging into Pulumi failed.\n{exc}\n{result}"
+            raise DataSafeHavenPulumiError(msg) from exc
+
+        print(result)
+        return typer.confirm("Is this the Pulumi account you expect?\n")
```

Comment:
There's a wrapper for this in logger (which means it will also get logged to whichever logging handlers are configured, not just written to stdout).